In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

# Módulo 07, Tarefa 01

Vamos começar a mexer na nossa base de projeto? Já fizemos um exercício de montar a variável resposta, já pudemos perceber que essa atividade pode não ser trivial. Vamos agora trabalhar a base para que fique propícia ao *scikitlearn* para trabalharmos.

Lembrando, a base se chama demo01.csv, e originalmente está publicada [aqui](https://www.kaggle.com/rikdifos/credit-card-approval-prediction).

#### 1) Carregue a base e avalie:

- As variáveis
- Tipos de dados de cada variável
- Quantidade de missings
- Distribuição da variável resposta (mau)

In [2]:
demo = pd.read_csv('demo01.csv')
demo.head()

sexo posse_de_veiculo posse_de_imovel  qtd_filhos            tipo_renda  \
0    M                Y               Y           0               Working   
1    F                N               Y           0  Commercial associate   
2    F                N               Y           0  Commercial associate   
3    M                Y               Y           0               Working   
4    F                Y               N           0               Working   

                        educacao          estado_civil    tipo_residencia  \
0  Secondary / secondary special               Married  House / apartment   
1  Secondary / secondary special  Single / not married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3               Higher education               Married  House / apartment   
4              Incomplete higher               Married  House / apartment   

       idade  tempo_emprego  possui_celular  possui_fone_comercial  \
0  58.832877       3.106849               1                      0   
1  52.356164       8.358904               1                      0   
2  52.356164       8.358904               1                      0   
3  46.224658       2.106849               1                      1   
4  29.230137       3.021918               1                      0   

   possui_fone  possui_email  qt_pessoas_residencia    mau  
0            0             0                    2.0  False  
1            1             1                    1.0  False  
2            1             1                    1.0  False  
3            1             1                    2.0  False  
4            0             0                    2.0  False

In [3]:
primeiro_shape = demo.shape

In [4]:
#Tipos de variáveis
demo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16650 entries, 0 to 16649
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   sexo                   16650 non-null  object 
 1   posse_de_veiculo       16650 non-null  object 
 2   posse_de_imovel        16650 non-null  object 
 3   qtd_filhos             16650 non-null  int64  
 4   tipo_renda             16650 non-null  object 
 5   educacao               16650 non-null  object 
 6   estado_civil           16650 non-null  object 
 7   tipo_residencia        16650 non-null  object 
 8   idade                  16650 non-null  float64
 9   tempo_emprego          16650 non-null  float64
 10  possui_celular         16650 non-null  int64  
 11  possui_fone_comercial  16650 non-null  int64  
 12  possui_fone            16650 non-null  int64  
 13  possui_email           16650 non-null  int64  
 14  qt_pessoas_residencia  16650 non-null  float64
 15  ma

Das 16 variáveis, tem-se: 1 booleana, 3 float, 1 inteiro e 11 são strigs

In [5]:
demo = demo.drop_duplicates()
segundo_shape=demo.shape

In [6]:
primeiro_shape

(16650, 16)

In [7]:
segundo_shape

(6770, 16)

Haviam 9880 dados duplicados. O arquivo demo, após retirada dos dados duplicados, agora possui 6770 dados

In [8]:
demo.isna().sum()

sexo                     0
posse_de_veiculo         0
posse_de_imovel          0
qtd_filhos               0
tipo_renda               0
educacao                 0
estado_civil             0
tipo_residencia          0
idade                    0
tempo_emprego            0
possui_celular           0
possui_fone_comercial    0
possui_fone              0
possui_email             0
qt_pessoas_residencia    0
mau                      0
dtype: int64

Não há dadso faltantes

In [9]:
demo['mau'].value_counts()

False    6472
True      298
Name: mau, dtype: int64

A variável mau, apresenta uma distribuição de 298 registros verdadeiros

#### 2) Vamos montar um metadados

1. Crie um dataframe com os nomes de cada variável e o tipo de dados de cada variável.



In [10]:
n_var = pd.DataFrame(pd.Series(dict(demo.dtypes)))
n_var

0
sexo                    object
posse_de_veiculo        object
posse_de_imovel         object
qtd_filhos               int64
tipo_renda              object
educacao                object
estado_civil            object
tipo_residencia         object
idade                  float64
tempo_emprego          float64
possui_celular           int64
possui_fone_comercial    int64
possui_fone              int64
possui_email             int64
qt_pessoas_residencia  float64
mau                       bool

In [11]:
n_var = n_var.reset_index()
n_var

index        0
0                    sexo   object
1        posse_de_veiculo   object
2         posse_de_imovel   object
3              qtd_filhos    int64
4              tipo_renda   object
5                educacao   object
6            estado_civil   object
7         tipo_residencia   object
8                   idade  float64
9           tempo_emprego  float64
10         possui_celular    int64
11  possui_fone_comercial    int64
12            possui_fone    int64
13           possui_email    int64
14  qt_pessoas_residencia  float64
15                    mau     bool

In [12]:
n_var.columns = ['Variavel', 'Tipo']
n_var

Variavel     Tipo
0                    sexo   object
1        posse_de_veiculo   object
2         posse_de_imovel   object
3              qtd_filhos    int64
4              tipo_renda   object
5                educacao   object
6            estado_civil   object
7         tipo_residencia   object
8                   idade  float64
9           tempo_emprego  float64
10         possui_celular    int64
11  possui_fone_comercial    int64
12            possui_fone    int64
13           possui_email    int64
14  qt_pessoas_residencia  float64
15                    mau     bool

2. Adicione uma coluna nesse *dataframe* chamada "qtd_categorias" e coloque nela o número de categorias correspondente de cada variável. 
    Dica: 
        1. inicie uma lista vazia
        2. faça um for ao longo dos nomes das variáveis, 
        3. conte o número de categorias dessa variável 
        4. acumule essa informação de 3. na lista que você criou em 1. 
        5. No final, essa lista pode ser atribuída à nossa variável.

In [13]:
nova_coluna = []
for i in demo.columns:
  nova_coluna.append(len(pd.unique(demo[i])))
nova_coluna 

[2, 2, 2, 8, 5, 5, 5, 6, 5298, 3005, 1, 2, 2, 2, 9, 2]

In [14]:
n_var['qtd_categorias'] = nova_coluna

In [15]:
n_var

Variavel     Tipo  qtd_categorias
0                    sexo   object               2
1        posse_de_veiculo   object               2
2         posse_de_imovel   object               2
3              qtd_filhos    int64               8
4              tipo_renda   object               5
5                educacao   object               5
6            estado_civil   object               5
7         tipo_residencia   object               6
8                   idade  float64            5298
9           tempo_emprego  float64            3005
10         possui_celular    int64               1
11  possui_fone_comercial    int64               2
12            possui_fone    int64               2
13           possui_email    int64               2
14  qt_pessoas_residencia  float64               9
15                    mau     bool               2

#### 3) Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}. Crie um *dataframe* apenas com as variáveis apropriadas para entrada no scikitlearn - elimine as variáveis tipo *str*, mantendo apenas suas versões *dummy*.

In [16]:
demo.sexo = demo.sexo.map({'F': 1, 'M': 0})
demo.posse_de_veiculo = demo.posse_de_veiculo.map({'Y': 1, 'N': 0})
demo.posse_de_imovel = demo.posse_de_imovel.map({'Y': 1, 'N': 0})
demo.mau = demo.mau.astype(int)

demo.head(5)

sexo  posse_de_veiculo  posse_de_imovel  qtd_filhos            tipo_renda  \
0     0                 1                1           0               Working   
1     1                 0                1           0  Commercial associate   
3     0                 1                1           0               Working   
4     1                 1                0           0               Working   
6     1                 0                1           0               Working   

                        educacao          estado_civil    tipo_residencia  \
0  Secondary / secondary special               Married  House / apartment   
1  Secondary / secondary special  Single / not married  House / apartment   
3               Higher education               Married  House / apartment   
4              Incomplete higher               Married  House / apartment   
6  Secondary / secondary special               Married  House / apartment   

       idade  tempo_emprego  possui_celular  possui_fone_comercial  \
0  58.832877       3.106849               1                      0   
1  52.356164       8.358904               1                      0   
3  46.224658       2.106849               1                      1   
4  29.230137       3.021918               1                      0   
6  27.482192       4.024658               1                      0   

   possui_fone  possui_email  qt_pessoas_residencia  mau  
0            0             0                    2.0    0  
1            1             1                    1.0    0  
3            1             1                    2.0    0  
4            0             0                    2.0    0  
6            1             0                    2.0    0

#### 4) Qual variável é mais poderosa?

Considere as variáveis ```possui_email``` e ```posse_de_veiculo```. Faça uma tabela cruzada entre elas e responda qual delas te parece mais poderosa para prever a probabilidade de ```mau = 1```?

In [27]:
demo[['mau','posse_de_veiculo','possui_email']].groupby(['mau']).count()

posse_de_veiculo  possui_email
mau                                
0                6472          6472
1                 298           298

In [19]:
demo[['mau','posse_de_veiculo','possui_email']].groupby(['mau','possui_email']).count()

posse_de_veiculo
mau possui_email                  
0   0                         5927
    1                          545
1   0                          278
    1                           20

In [20]:
demo[['mau','posse_de_veiculo','possui_email']].groupby(['mau','posse_de_veiculo']).count()

possui_email
mau posse_de_veiculo              
0   0                         4021
    1                         2451
1   0                          190
    1                          108

posse_de_veiculo tem é mais poderosa para prever a probabilidade de mau = 1?

#### 5) Salve a base, pois ela será utilizada no final deste módulo.

In [28]:
demo.to_csv('demo01_tratado.csv')